### DISCUSSION TITLE
Tokio timeout stuck on a loop.

### DISCUSSION THREAD

**Discussion by galeaspablo** [tokio-rs/tokio#4821] [Q&A] [answered]
This is a simplified version of my code,
```
async fn empty_await() -> i32 {
    0
}

#[tokio::main(flavor="multi_thread", worker_threads = 6)]
async fn main() {
    let this_should_timeout_but_does_not = tokio::time::timeout(
        Duration::from_millis(1),
        async move {
            let mut x = 0;
            loop {
                empty_await().await;
                x += 1;
                println!("x = {}", x);
            }
        }
    ).await.unwrap();

    println!("line never reached");
}
```

The output of this program is

```
....
x = 6533492
x = 6533493
x = 6533494
x = 6533495
x = 6533496
x = 6533497
x = 6533498
x = 6533499
x = 6533500
```

It will keep going until I kill the process. Do I fundamentally misunderstand tokio and async? Isn't the timeout meant to stop the infinite loop? I read [here](https://ryhl.io/blog/async-what-is-blocking/) that calling await gives the runtime a chance to swap tasks (hence the `empty_await`).

**Comment by galeaspablo** [comment]
This one panics as expected. How come the other one doesn't?

```
#[tokio::main(flavor="multi_thread", worker_threads = 6)]
async fn main() {
    let this_should_timeout_but_does_not = tokio::time::timeout(
        Duration::from_millis(1),
        async move {
            let mut x = 0;
            loop {
                tokio::time::sleep(Duration::from_millis(5)).await;
                x += 1;
                println!("x = {}", x);
            }
        }
    ).await.unwrap();

    println!("line never reached");
}
```

**Comment by Noah-Kennedy** [comment, marked as answer]
So, the future created by the `async move` block never actually yields because the inner future you await is always ready.

Rewrite empty wait as such:
```rust
async fn empty_await() -> i32 {
    tokio::task::yield_now().await;
    0
}
```

Many tokio async functions and futures use an internal budgeting system to force futures which would otherwise be ready to occasionally yield in order to mitigate this issue.

**Reply by galeaspablo** [reply]
Makes sense. Works as expected now :) Thank you

**Reply by Noah-Kennedy** [reply]
No problem!
